### PR TITLE
`FRAX` series proposal

### PR DESCRIPTION
**Pull request by iamsahu**
# Proposal

Introduce FRAX as a borrowable asset.

# Background

You can find more details [here](https://gov.frax.finance/t/yield-protocol-frax-proposal-call-for-comments/835)

# Details

The summary of the execution is as follows:

1. Deploy Accumulator Oracle (no need for governance approval)
2. Deploy FRAX Join (no need for governance approval)
3. Deploy the FYFRAX2209 and FYFRAX2303 fyToken & pools (no need for governance approval)
4. Deploy YSFRAX1YEV and YSFRAX1YOD strategies (no need for governance approval)
5. Deploy OnChainTest (no need for governance approval)
6. Governance proposal
   a. Update chi and rate in accumulator oracle
   b. Update Chainlink sources
   c. Update sources in the Composite Oracle
   d. Update Paths in the Composite Oracle
   e. Add FRAX as an asset
   f. Make FRAX a base
   g. Approve DAI, USDC, WBTC, WSTETH, ENS and LINK as collaterals for FRAX using the Composite Oracle
   h. Approve ETH as collateral for FRAX using the Chainlink Oracle
   i. Add the FYFRAX2209 and FYFRAX2303 fyToken and pools as series
   j. Initialize the FYETH2206 and FYETH2209 pools with 100 FRAX each, and burn the LP tokens
   k. Orchestrate the YSETH6MMS and YSETH6MJD strategies
   l. Initialize the YSETH6MMS and YSETH6MJD strategies with 100 FRAX each, and burn the strategy tokens

These steps in 6 are encoded into the [addFraxSeries.ts](https://github.com/yieldprotocol/environments-v2/blob/aeb8eb9d8f45052103f0514adef129616c27ebea/scripts/governance/addSeries/addFraxSeries/addFraxSeries.ts) scripts, while the steps 1 to 5 are the regular deployment scripts from the [environment deployment](https://github.com/yieldprotocol/environments-v2/tree/feat/eth-series/scripts/governance/redeploy).

[This input](https://github.com/yieldprotocol/environments-v2/blob/aeb8eb9d8f45052103f0514adef129616c27ebea/scripts/governance/addSeries/addFraxSeries/addFraxSeries.mainnet.config.ts) will be used:

## Proposal Config

```
/// @notice Configuration of the acummulator
/// @param Asset identifier (bytes6 tag)
/// @param Acummulator type (bytes6 tag)
/// @param start Initial value for the acummulator (18 decimal fixed point)
/// @param increasePerSecond Acummulator multiplier, per second (18 decimal fixed point)
export const rateChiSources: Array<[string, string, string, string]> = [
  [FRAX, RATE, WAD.toString(), '1000000001546067000'],
  [FRAX, CHI, WAD.toString(), WAD.toString()],
]

/// @notice Assets that will be added to the protocol
/// @param Asset identifier (bytes6 tag)
/// @param Address for the asset
export const assetsToAdd: Map<string, string> = new Map([[FRAX, assets.get(FRAX) as string]])

/// @notice Assets that will be made into a base
/// @param Asset identifier (bytes6 tag)
/// @param Address for the related Join
export const bases: Array<[string, string]> = [[FRAX, newJoins.get(FRAX) as string]]

/// @notice Sources that will be added to the Chainlink Oracle
/// @param Base asset identifier (bytes6 tag)
/// @param Address for the base asset
/// @param Quote asset identifier (bytes6 tag)
/// @param Address for the quote asset
/// @param Address for the chainlink aggregator
export const chainlinkSources: Array<[string, string, string, string, string]> = [
  [FRAX, assets.get(FRAX) as string, ETH, assets.get(ETH) as string, '0x14d04Fff8D21bd62987a5cE9ce543d2F1edF5D3E'],
]

/// @notice Oracles that will be used for Chi
/// @param Asset identifier (bytes6 tag)
/// @param Address for the chi oracle
export const newChiSources: Array<[string, string]> = [[FRAX, protocol.get(ACCUMULATOR) as string]]

/// @notice Oracles that will be used for Rate
/// @param Asset identifier (bytes6 tag)
/// @param Address for the rate oracle
export const newRateSources: Array<[string, string]> = [[FRAX, protocol.get(ACCUMULATOR) as string]]

/// @notice Sources that will be added to the Composite Oracle
/// @param Base asset identifier (bytes6 tag)
/// @param Quote asset identifier (bytes6 tag)
/// @param Address for the source
export const compositeSources: Array<[string, string, string]> = [[FRAX, ETH, protocol.get(CHAINLINK) as string]]

/// @notice Paths that will be added to the Composite Oracle
/// @param Base asset identifier (bytes6 tag)
/// @param Quote asset identifier (bytes6 tag)
/// @param Path to traverse (array of bytes6 tags)
export const newCompositePaths: Array<[string, string, Array<string>]> = [
  [FRAX, ENS, [ETH]],
  [FRAX, WSTETH, [ETH, STETH]],
]

/// @notice Configure an asset as an ilk for a base using the Chainlink Oracle
/// @param Base asset identifier (bytes6 tag)
/// @param Ilk asset identifier (bytes6 tag)
/// @param Collateralization ratio as a fixed point number with 6 decimals
/// @param Debt ceiling, modified by decimals
/// @param Minimum vault debt, modified by decimals
/// @param Decimals to append to debt ceiling and minimum vault debt.
export const newChainlinkLimits: Array<[string, string, number, number, number, number]> = [
  [FRAX, FRAX, 1000000, 5000000, 5000, 18],
  [FRAX, ETH, 1400000, 1000000, 5000, 18],
  [FRAX, DAI, 1100000, 1000000, 5000, 18],
  [FRAX, USDC, 1100000, 1000000, 5000, 18],
  [FRAX, WBTC, 1500000, 1000000, 5000, 18],
  [FRAX, LINK, 1670000, 1000000, 5000, 18],
  [FRAX, UNI, 1670000, 1000000, 5000, 18],
  [DAI, FRAX, 1100000, 100000, 5000, 18],
  [USDC, FRAX, 1100000, 100000, 5000, 6],
  [ETH, FRAX, 1400000, 1000000, 5000, 12],
]

/// @notice Configure an asset as an ilk for a base using the Composite Oracle
/// @param Base asset identifier (bytes6 tag)
/// @param Ilk asset identifier (bytes6 tag)
/// @param Collateralization ratio as a fixed point number with 6 decimals
/// @param Debt ceiling, modified by decimals
/// @param Minimum vault debt, modified by decimals
/// @param Decimals to append to debt ceiling and minimum vault debt.
export const newCompositeLimits: Array<[string, string, number, number, number, number]> = [
  [FRAX, WSTETH, 1400000, 1000000, 5000, 18],
  [FRAX, ENS, 1670000, 1000000, 5000, 18],
]

/// @notice Deploy fyToken series
/// @param series identifier (bytes6 tag)
/// @param underlying identifier (bytes6 tag)
/// @param Address for the chi oracle
/// @param Address for the related Join
/// @param Maturity in unix time (seconds since Jan 1, 1970)
/// @param Name for the series
/// @param Symbol for the series
export const fyTokenData: Array<[string, string, string, string, number, string, string]> = [
  [
    FYFRAX2206,
    FRAX,
    protocol.get(ACCUMULATOR) as string,
    newJoins.get(FRAX) as string,
    EOJUN22,
    'FYFRAX2206',
    'FYFRAX2206',
  ],
  [
    FYFRAX2209,
    FRAX,
    protocol.get(ACCUMULATOR) as string,
    newJoins.get(FRAX) as string,
    EOSEP22,
    'FYFRAX2209',
    'FYFRAX2209',
  ],
]

/// @notice Deploy YieldSpace pools
/// @param pool identifier, usually matching the series (bytes6 tag)
/// @param base address
/// @param fyToken address
/// @param time stretch, in 64.64
/// @param g1, in 64.64
/// @param g2, in 64.64
export const poolData: Array<[string, string, string, BigNumber, BigNumber, BigNumber]> = [
  [
    FYFRAX2206,
    assets.get(FRAX) as string,
    newFYTokens.get(FYFRAX2206) as string,
    ONE64.div(secondsInOneYear.mul(37)),
    ONE64.mul(80).div(100),
    ONE64.mul(100).div(80),
  ],
  [
    FYFRAX2209,
    assets.get(FRAX) as string,
    newFYTokens.get(FYFRAX2209) as string,
    ONE64.div(secondsInOneYear.mul(37)),
    ONE64.mul(80).div(100),
    ONE64.mul(100).div(80),
  ],
]

/// @notice Pool initialization parameters
/// @param pool identifier, usually matching the series (bytes6 tag)
/// @param base identifier (bytes6 tag)
/// @param amount of base to initialize pool with
/// @param amount of fyToken to initialize pool with
export const poolsInit: Array<[string, string, BigNumber, BigNumber]> = [
  [FYFRAX2206, FRAX, WAD.mul(100), ZERO],
  [FYFRAX2209, FRAX, WAD.mul(100), ZERO],
]

/// @notice Ilks to accept for series
/// @param series identifier (bytes6 tag)
/// @param newly accepted ilks (array of bytes6 tags)
export const seriesIlks: Array<[string, string[]]> = [
  [FYFRAX2206, [FRAX, ETH, DAI, USDC, WBTC, WSTETH, LINK, ENS, UNI]],
  [FYFRAX2209, [FRAX, ETH, DAI, USDC, WBTC, WSTETH, LINK, ENS, UNI]],
  [FYETH2206, [FRAX]],
  [FYETH2209, [FRAX]],
  [FYDAI2206, [FRAX]],
  [FYDAI2209, [FRAX]],
  [FYUSDC2206, [FRAX]],
  [FYUSDC2209, [FRAX]],
]

/// @notice Deploy strategies
/// @param strategy name
/// @param strategy identifier (bytes6 tag)
/// @param Address for the related Join
/// @param Address for the related asset
export const strategiesData: Array<[string, string, string, string, string]> = [
  // name, symbol, baseId
  ['Yield Strategy FRAX 6M Mar Sep', YSFRAX6MMS, FRAX, newJoins.get(FRAX) as string, assets.get(FRAX) as string],
  ['Yield Strategy FRAX 6M Jun Dec', YSFRAX6MJD, FRAX, newJoins.get(FRAX) as string, assets.get(FRAX) as string],
]

/// @notice Strategy initialization parameters
/// @param strategy identifier (bytes6 tag)
/// @param address of initial pool
/// @param series identifier (bytes6 tag)
/// @param amount of base to initialize strategy with
export const strategiesInit: Array<[string, string, string, BigNumber]> = [
  // [strategyId, startPoolAddress, startPoolId, initAmount]
  [YSFRAX6MMS, newPools.get(FYFRAX2209) as string, FYFRAX2209, WAD.mul(100)],
  [YSFRAX6MJD, newPools.get(FYFRAX2206) as string, FYFRAX2206, WAD.mul(100)],
]

// Input data: ilkId, duration, initialOffer, auctionLine, auctionDust, dec
/// @notice Limits to be used in an auction
/// @param base identifier (bytes6 tag)
/// @param duration of auctions in seconds
/// @param initial percentage of the collateral to be offered (fixed point with 6 decimals)
/// @param Maximum concurrently auctionable for this asset, modified by decimals
/// @param Minimum vault debt, modified by decimals
/// @param Decimals to append to auction ceiling and minimum vault debt.
export const chainlinkAuctionLimits: Array<[string, number, number, number, number, number]> = [
  [FRAX, 3600, 1000000, 1000000, 5000, 18],
]
```

# Testing

The series was tested by running [this test](https://github.com/yieldprotocol/environments-v2/blob/aeb8eb9d8f45052103f0514adef129616c27ebea/scripts/governance/addSeries/addFraxSeries/addFraxSeries.test.ts)

Here is the result:

```
+ npx hardhat run --network localhost ./scripts/governance/addSeries/addFraxSeries/addFraxSeries.test.ts
No need to generate any newer typings.
ChainId: 4
Running on a fork, impersonating 0x5AD7799f02D5a829B2d6FA085e6bd69A872619D5
series: 0x303330360000
ilk: 0x313800000000
vault: 0x65a6a08fd91ab5e635b96030
posting 5050000000000000000000 ilk out of 106988301411298813028965763
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 5050000000000000000000 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x303000000000
vault: 0x76ae5112c92fb07fa11bf722
posting 2321633476618397690 ilk out of 93299393276207498942111
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 2321633476618397690 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x303100000000
vault: 0xac4e851502351c9b5c6b2cae
posting 6716011343812344723202 ilk out of 96286893322197418742174162
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 6716011343812344723202 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x303200000000
vault: 0x0ac9c03cd8315c7060a732d5
posting 6722036672 ilk out of 398655589693586
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 6722036672 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x303300000000
vault: 0x057e6b737bab7381e95c3d44
posting 18725119 ilk out of 713243120670
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 18725119 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x303400000000
vault: 0x2b67829f238d19014534723c
posting 2175955139545211497 ilk out of 166033565564884889141328
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 2175955139545211497 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x303600000000
vault: 0x9b445f5bf44a415a6720ccb1
posting 600405834621211816167 ilk out of 12805409374694245207358465
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 600405834621211816167 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x303700000000
vault: 0x874ed212d7d55248764fb4d9
posting 588663348274580087426 ilk out of 65532017200000000000000000
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 588663348274580087426 ilk
repaid and withdrawn
series: 0x303330360000
ilk: 0x313000000000
vault: 0x0679199c0c66c1f67c710fb7
posting 906708755663085073525 ilk out of 24679867136918000000000000
borrowing 5000000000000000000000 FYFRAX2206
posted and borrowed
repaying 5000000000000000000000 FYFRAX2206 and withdrawing 906708755663085073525 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x313800000000
vault: 0x040c8a151e1952180a9fa717
posting 5050000000000000000000 ilk out of 106988301411298813028965763
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 5050000000000000000000 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x303000000000
vault: 0x96f86e027930f15cd1211f2e
posting 2321633476618397690 ilk out of 93299393276207498942111
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 2321633476618397690 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x303100000000
vault: 0xa06292745cd614323f9d04e5
posting 6716011343812344723202 ilk out of 96286893322197418742174162
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 6716011343812344723202 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x303200000000
vault: 0x797b590801aa4d54ee6a596a
posting 6722036672 ilk out of 398655589693586
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 6722036672 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x303300000000
vault: 0x210514238a6887603c9615d2
posting 18725119 ilk out of 713243120670
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 18725119 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x303400000000
vault: 0xcebfa63ee167979e6ea040c5
posting 2175955139545211497 ilk out of 166033565564884889141328
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 2175955139545211497 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x303600000000
vault: 0x787f79cb284efad96d27039a
posting 600405834621211816167 ilk out of 12805409374694245207358465
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 600405834621211816167 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x303700000000
vault: 0x0e83a23caae38410603f2d0a
posting 588663348274580087426 ilk out of 65532017200000000000000000
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 588663348274580087426 ilk
repaid and withdrawn
series: 0x303330370000
ilk: 0x313000000000
vault: 0x5aab9a41eeeb68a753e8e38a
posting 906708755663085073525 ilk out of 24679867136918000000000000
borrowing 5000000000000000000000 FYFRAX2209
posted and borrowed
repaying 5000000000000000000000 FYFRAX2209 and withdrawing 906708755663085073525 ilk
repaid and withdrawn
series: 0x303030360000
ilk: 0x313800000000
vault: 0xe2f91e6198f65b47c2435741
posting 46135835427396165759 ilk out of 106988301411298813028965763
borrowing 10000000000000000 FYETH2206
posted and borrowed
repaying 10000000000000000 FYETH2206 and withdrawing 46135835427396165759 ilk
repaid and withdrawn
series: 0x303030370000
ilk: 0x313800000000
vault: 0xbb60d7a77e82b739d3926690
posting 46135835427396165759 ilk out of 106988301411298813028965763
borrowing 10000000000000000 FYETH2209
posted and borrowed
repaying 10000000000000000 FYETH2209 and withdrawing 46135835427396165759 ilk
repaid and withdrawn
series: 0x303130360000
ilk: 0x313800000000
vault: 0xae8eb72618a2b34cc8a4dfcd
posting 6716988691742221448543 ilk out of 106988301411298813028965763
borrowing 5000000000000000000000 FYDAI2206
posted and borrowed
repaying 5000000000000000000000 FYDAI2206 and withdrawing 6716988691742221448543 ilk
repaid and withdrawn
series: 0x303130370000
ilk: 0x313800000000
vault: 0xa8128ebb59318bea7aa2979e
posting 6716988691742221448543 ilk out of 106988301411298813028965763
borrowing 5000000000000000000000 FYDAI2209
posted and borrowed
repaying 5000000000000000000000 FYDAI2209 and withdrawing 6716988691742221448543 ilk
repaid and withdrawn
series: 0x303230360000
ilk: 0x313800000000
vault: 0x905f0dd5c3518b2124b5983f
posting 6710967886837505321759 ilk out of 106988301411298813028965763
borrowing 5000000000 FYUSDC2206
posted and borrowed
repaying 5000000000 FYUSDC2206 and withdrawing 6710967886837505321759 ilk
repaid and withdrawn
series: 0x303230370000
ilk: 0x313800000000
vault: 0x55aa491070f59b7f33c29233
posting 6710967886837505321759 ilk out of 106988301411298813028965763
borrowing 5000000000 FYUSDC2209
posted and borrowed
repaying 5000000000 FYUSDC2209 and withdrawing 6710967886837505321759 ilk
repaid and withdrawn
```

### CODE REVIEW

File: YPP-0034.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+export const newCompositePaths: Array<[string, string, Array<string>]> = [
+  [FRAX, ENS, [ETH]],
+  [FRAX, WSTETH, [ETH, STETH]],
+]
+
+// Input data: baseId, ilkId, ratio (1000000 == 100%), line, dust, dec
+export const newChainlinkLimits: Array<[string, string, number, number, number, number]> = [
+  [FRAX, FRAX, 1000000, 5000000, 5000, 18],
+  [FRAX, ETH, 1400000, 1000000, 5000, 18],
+  [FRAX, DAI, 1330000, 1000000, 5000, 18],
```
Comment (aniemerg) [inline YPP-0034.md:70]: For FRAX/DAI and FRAX/USD, we can probably be way less strict on collateralization ratio. Let's do 110%

**Comment by alcueca** [comment]
Proposed as `0xcce76baacdbea977e53dc6232bf5f9b7dd58480a93ce195c8ca8aa5c343fb3c4`

**Review by alcueca — APPROVED** [review]
LGTM

**Comment by alcueca** [comment]
[Executed](https://etherscan.io/tx/0x7bf10a90fb8ea4902e144f9702dddbc040b5c72cebd96155d076c318dfff9cf5)

**Pull request merged**